_posts/2020-06-10-application-components.adoc

```diff
@@ -0,0 +1,574 @@
+  </layout>
+</window>
+----
+<1> a input dialog facet is used for requesting the visit type before the visit is created
+<2> by redefining the `groupTable` with the same ID as in the parent screen: "petsTable", the merging of the two screen descriptors is initiated
+<3> the button i    s appended to the existing `buttonsPanel` component
+
+The business logic of the screen is also extended accordingly. The controller subscribes to the close event of the input dialog.
+
+
+.KantoPetBrowse.java
+[source,java]
+----
+@UiController("petclinic_Pet.browse")
+@UiDescriptor("kanto-pet-browse.xml")
+public class KantoPetBrowse extends PetBrowse {
+
+    @Inject
+    protected GroupTable<Pet> petsTable;
+    @Inject
+    protected Notifications notifications;
+    @Inject
+    protected VisitCreationService visitCreationService;
+    @Inject
+    protected ScreenBuilders screenBuilders;
+    @Inject
+    protected MessageBundle messageBundle;
+    @Inject
+    protected NotificationFacet visitCreatedNotification;
+
+    @Subscribe("createVisitForPetDialog")
+    protected void onCreateVisitForPetDialogClose(CloseEvent event) {
+        if (event.getCloseAction().equals(InputDialog.INPUT_DIALOG_OK_ACTION)) {
+            final VisitType visitType = (VisitType) event.getValues().get("visitType");
+
+            final Visit createdVisit = visitCreationService
+                .createVisitForPet(petsTable.getSingleSelected(), visitType); // <1>
+
+            visitCreatedNotification.show();
+
+            screenBuilders.editor(Visit.class, this)
+                .editEntity(createdVisit)
+                .withOpenMode(OpenMode.DIALOG)
+                .show();
+
+        }
+    }
+}
+----
+<1> the creation of the Visit is delegated to the `VisitCreationService`, which is only part of the Kanto-Petclinic project.
+
+The `VisitCreationService` is a new CUBA Service that is not part of the original Petclinic platform. Instead, we create this service exclusively for the Kanto-Petclinic project:
+
+
+.VisitCreationServiceBean.java
+[source,java]
+----
+
+@Service(VisitCreationService.NAME)
+public class VisitCreationServiceBean implements VisitCreationService {
+
+    @Inject
+    protected DataManager dataManager;
+    @Inject
+    protected TimeSource timeSource;
+
+    @Override
+    public Visit createVisitForPet(Pet pet, VisitType visitType) {
+
+        final Visit visitForPet = dataManager.create(Visit.class);
+
+        visitForPet.setPet(pet);
+        visitForPet.setType(visitType);
+
+        final LocalDateTime now = timeSource.now().toLocalDateTime();
+        visitForPet.setVisitStart(now);
+        visitForPet.setVisitEnd(now.plusHours(1));
+
+        return dataManager.commit(visitForPet);
+    }
+}
+----
+
+With this code in place, the customizations for the Kanto-Petclinic are finished. The application can now be shipped to the customer. Let's look at the second customer "Alabastia-Petclinic" so see some other extensions in action.
+
```

Comment:
Maybe show some screenshots of vanilla Petclinic vs the extension? I think it would simplify understanding of what can be achieved.

Reply:
yes, the screenshots are currently missing as there is a problem with the theme, that I'm currently trying to figure with @web-devel. When this is done - the plan was to create a couple of screenshots and show all the extensions.